src/khepri_fun.erl

```diff
@@ -330,55 +346,39 @@ add_comment_and_retry(
-  Asm, Error,
   [#function{name = Name, arity = Arity, code = Code} = Function | Rest],
-  Name, Arity, Comment, Result) ->
-    Code1 = add_comment_to_code(Asm, Error, Code, Comment, []),
+  {_GeneratedModuleName, Name, Arity},
+  Location, Comment, Result) ->
+    Code1 = add_comment_to_code(Code, Location, Comment),
     Function1 = Function#function{code = Code1},
     lists:reverse(Result) ++ [Function1 | Rest];
 add_comment_to_function(
-  Asm, Error,
-  [Function | Rest], Name, Arity, Comment, Result) ->
+  [#function{entry = EntryLabel, code = Code} = Function | Rest],
+  EntryLabel, Location, Comment, Result) ->
+    Code1 = add_comment_to_code(Code, Location, Comment),
+    Function1 = Function#function{code = Code1},
+    lists:reverse(Result) ++ [Function1 | Rest];
+add_comment_to_function(
+  [Function | Rest], FailingFun, Location, Comment, Result) ->
     add_comment_to_function(
-      Asm, Error, Rest, Name, Arity, Comment, [Function | Result]).
-
-add_comment_to_code(
-  Asm, Error,
-  [{label, _} = Instruction | Rest],
-  Comment, Result) ->
-    add_comment_to_code(Asm, Error, Rest, Comment, [Instruction | Result]);
-add_comment_to_code(
-  Asm, Error,
-  [{func_info, _, _, _} = Instruction | Rest],
-  Comment, Result) ->
-    add_comment_to_code(Asm, Error, Rest, Comment, [Instruction | Result]);
-add_comment_to_code(
-  Asm, Error,
-  [{'%', _} = Instruction | Rest],
-  Comment, Result) ->
-    case are_comments_conflicting(Instruction, Comment) of
-        false ->
-            add_comment_to_code(
-              Asm, Error, Rest, Comment, [Instruction | Result]);
-        true ->
-            throw(
-              {conflicting_assembly_annotations,
-               Instruction, Comment, Error, Asm})
-    end;
-add_comment_to_code(
-  _Asm, _Error,
-  Rest,
-  Comment, Result) ->
-    lists:reverse(Result) ++ [Comment | Rest].
-
-are_comments_conflicting(
-  {'%', {var_info, Register, _}},
-  {'%', {var_info, Register, _}}) ->
-    %% If we are about to generate two `var_info' comments affecting the same
-    %% register (i.e. same variable), we abort.
-    true;
-are_comments_conflicting(_Comment1, _Comment2) ->
-    false.
+      Rest, FailingFun, Location, Comment, [Function | Result]).
+
+add_comment_to_code(Code, Location, Comment) ->
+    %% The instruction counter from `beam_validator' is 1-indexed.
+    add_comment_to_code(Code, Location, Comment, 1, []).
+
+add_comment_to_code(Rest, Location, Comment, Location, Result) ->
+    Comment1 = merge_comments(Comment, hd(Result)),
+    lists:reverse(Result) ++ [Comment1 | Rest];
+add_comment_to_code([Instruction | Rest], Location, Comment, Counter, Result) ->
+    add_comment_to_code(
+        Rest, Location, Comment, Counter + 1, [Instruction | Result]).
+
+merge_comments(
+  {'%', {var_info, Var, Attributes1}},
+  {'%', {var_info, Var, Attributes2}}) ->
+    {'%', {var_info, Var, Attributes1 ++ Attributes2}};
+merge_comments(Comment, _Instruction) -> Comment.
```

Comment:
I was wrong about the merging behavior of mulitple comments, we do need to do the merging ourselves.

Reply:
I think this is a good idea but I don't think clauses for merging types would end up being used. For example with a pair that has its first and second element accessed, currently we add the annotations:

```erl
{'%',{var_info,{x,0}, [{type,t_tuple,1,false,#{}}]}}.
{get_tuple_element,{x,0},0,{y,2}}.
{'%',{var_info,{x,0}, [{type,t_tuple,2,false,#{}}]}}.
{get_tuple_element,{x,0},1,{x,0}}.
```

In the clauses added to pass1_process_instructions/3.

We could try to merge these annotations together but it seems complicated: the annotation will only be correct within the scope (all instructions between `{label, _}`s) and I'm not sure how it works when variables are moved around (for example `{move, {x, 1}, {x, 0}}`).

If we do run into cases where the validator disagrees with the types, I think it might make sense to have `merge_comments/2` replace the `{type_}` attribute in Attributes2 with the one in Attributes1 (prefer the one beam_validator says is `needed`)

Reply:
Ok, let's wait and revisit this if we hit a problematic situation.